# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [14]:
data.duplicated().sum()

54

In [15]:
data = data.drop_duplicates()

**Задание 15. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [16]:
data['education'] = data['education'].str.lower()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследуйте данные и ответьте на вопросы

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Определим какое количество должников присутствует в каждой группе заемщиков в разрезе количества детей.

In [22]:
data_children_debt = data.groupby('children').agg({'debt':['sum', 'count']})
#дополнительно рассчитаем процент должников для каждой группы
data_children_debt[('debt', 'debt_percentage')] = round(data_children_debt['debt']['sum']/data_children_debt['debt']['count']*100, 2)
data_children_debt

debt                       
           sum  count debt_percentage
children                             
0         1063  14107            7.54
1          444   4809            9.23
2          194   2052            9.45
3           27    330            8.18
4            4     41            9.76
5            0      9            0.00

Среди тех, у кого нет детей, имеющих задолженность по кредиту больше всего. Тем не менее процент должников в этой группе самый низкий, т.е. большое количество должников объясняется тем, что группа заемщиков без детей - самая большая по численности.

В группе заемщиков, у кого 5 детей, ни у кого нет задолженностей по кредиту. При этом в этой группе всего 9 заемщиков, т.е. данных недостаточно, чтобы формировать какой-то вывод по данной группе.

**Вывод:** 

Полученные результаты позволяют сделать следующие выводы:
- Доля должников среди заемщиков не зависит от количества детей, т.к. в каждой группе процент должников отличается незначительно (от 0,5-2%).

- Отсутствие должников в группе заемщиков с 5 детьми объясняется очень маленьким размером этой группы - всего 9  человек. Появление хотя бы 1 заемщика с задолженностью приведет сразу к тем же показателям, что и в группах с другим количеством детей (~10% должников).

Гипотеза о том, что есть зависимость между количеством детей и возвратом кредита в срок не подтвердилась.

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Определим какое количество должников присутствует в каждой группе заемщиков в разрезе семейного положения.

In [23]:
#в задании 19 использовалась групповая функция, в задания 20 решаю через pivot_table для разнообразия
data_family_debt = data.pivot_table(index='family_status',values='debt',aggfunc=['sum','count'])
data_family_debt[('debt_percentage', 'debt')] = round(data_family_debt[('sum', 'debt')]/data_family_debt[('count', 'debt')]*100, 2)
data_family_debt.sort_values(by = ('debt_percentage', 'debt'))

sum  count debt_percentage
                      debt   debt            debt
family_status                                    
вдовец / вдова          63    951            6.62
в разводе               84   1189            7.06
женат / замужем        927  12266            7.56
гражданский брак       385   4146            9.29
Не женат / не замужем  273   2796            9.76

**Вывод:** 

Чаще всего задолженность возникает у заемщиков, не состоящих в официальных отношениях (статусы: Не женат / не замужем, гражданский брак). 

Гипотеза подтвердилась.

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

Определим какое количество должников присутствует в каждой группе заемщиков в разрезе уровня дохода.

In [24]:
data_income_debt = data.pivot_table(index='total_income_category',values='debt',aggfunc=['sum','count'])
data_income_debt[('debt_percentage', 'debt')] = round(data_income_debt[('sum', 'debt')]/data_income_debt[('count', 'debt')]*100, 2)
data_income_debt

sum  count debt_percentage
                       debt   debt            debt
total_income_category                             
A                         2     25            8.00
B                       354   5014            7.06
C                      1353  15938            8.49
D                        21    349            6.02
E                         2     22            9.09

**Вывод:** 

Зависимость между уровнем дохода и возвратом кредита в срок не обнаружена. Даже если пренебречь незначительным количеством заемщиков в категориях E (доход до 30000) и A (доход более 1 млн), то все равно явную зависимость не выделить.  
Гипотеза не подтвердилась.

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

Определим какое количество должников присутствует в каждой группе заемщиков в разрезе целей кредитования.

In [25]:
data_purpose_debt = data.pivot_table(index='purpose_category',values='debt',aggfunc=['sum','count'])
data_purpose_debt[('debt_percentage', 'debt')] = round(data_purpose_debt[('sum', 'debt')]/data_purpose_debt[('count', 'debt')]*100, 2)
data_purpose_debt.sort_values(by = ('debt_percentage', 'debt'))

sum  count debt_percentage
                         debt   debt            debt
purpose_category                                    
операции с недвижимостью  780  10754            7.25
проведение свадьбы        183   2324            7.87
получение образования     369   3989            9.25
операции с автомобилем    400   4281            9.34

**Вывод:** 

Наибольшее количество просроченной задолженности возникает по кредитам, выданным на операции с автомобилями (9,34%) и на получение образования (9,25%).

**Задание 23. Приведите возможные причины появления пропусков в исходных данных.**

*Ответ:*   
В исходных данных пропуски были в 2-х столбцах:  
*days_employed (общий трудовой стаж в дня)       2174*  
*total_income (ежемесячный доход)       2174*  

Проверка пустых значений показала, что пропуски в полях days_employed и total_income присутствуют на одних и тех же строках.
Это позволяет предположить технологическую ошибку в данных, т.к. маловероятно, что человек мог ошибиться сразу в двух столбцах и на одних и тех же строках.

Тем не менее, полностью нельзя исключать вероятность ошибки из-за человеческого фактора. Например, в столбце 'education' уровень образования клиента заполнен в совершенно разных регистрах, т.е. это допускает возможность внесения некорретных данных пользователями.

**Задание 24. Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.**

*Ответ:*   
Заполнять пропуски в количественных данных медианным значением является лучшим решением, т.к. данное значение не приводит к аномалиям в данных и искажению исследования.

Использовать среднее арифметическое для этих целей будет ошибкой из-за возможных выбросов в данных, из-за которых "среднее" на самом деле не будет отражать усредненное состояние на основе большинства данных. 

Так же для количественных данных обычно нельзя выделить некоторое значение по умолчанию, т.к. это может исказить результаты исследования. Например, если в текущей задаче все пропуски в столбце total_income заменили бы на 0, то проверка зависимости уровня дохода от возврата кредита могла бы привести к неожиданным некорректным результатам.

## Общий вывод.

Для Заказчика (кредитного отдела банка) было проведено исследование платежеспособности клиентов для последующего учета результатов при построении модели кредитного скоринга.

Были проверены следующие гипотезы:
1. Наличие зависимости между количеством детей и возвратом кредита в срок.  
Гипотеза не подтвердилась.
2. Наличие зависимости между семейным положением и возвратом кредита в срок.  
Гипотеза подтвердилась. Кредиты, выданные заемщикам, не состоящим в официальных отношениях (статусы: Не женат / не замужем, гражданский брак), чаще попадают в просрочку.
3. Наличие зависимости между уровнем дохода и возвратом кредита в срок.  
Гипотеза не подтвердилась.

В результате анализа целей кредитования было выявлено, что кредиты, выданные на операции с автомобилями и на получение образования, чаще попадают в просрочку, чем остальные.

При построении модели кредитного скоринга следует обратить внимание на следующие факторы:
- Семейное положение заемщиков
- Цели кредита (операции с автомобилями, получение образования)


Дополнительно были приведены общие выводы о возможных причинах появления пропусков в данных и дано обоснование выбора значений, которыми были заполнены данные пропуски.